### PR TITLE
Update MF2 definition to 2024-01-24 version

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
- Lexer ✅ 
- AST/Parser ✅ 
- Builder ❌

### CODE REVIEW

File: parse/lex.go [hunk 1]
```diff
@@ -508,7 +523,22 @@ func lexIdentifier(l *lexer) stateFn {
 			l.backup()
 
 			return l.emitItem(mk(typ, s))
-		case len(s) > 0 && isName(r):
+		case typ == 0:
```
Comment (kaaposc) [inline parse/lex.go:526]: why not `itemError`?

Comment (kaaposc) [reply parse/lex.go:526]: what if rename `itemError` to `itemNone`?

Comment (VladislavsPerkanuks) [reply parse/lex.go:526]: we need itemError, it is used to handle errors.

Comment (VladislavsPerkanuks) [reply parse/lex.go:526]: but i will change swithc to itemError with comment, that that is an zero value for item type

Comment (kaaposc) [reply parse/lex.go:526]: `itemNone` would imply an error since items has to have some type 🤔
anyway treat this just as a suggestion for alternative

**Comment by janishorsts** [comment]
@VladislavsPerkanuks can you please provide diff screenshot? So it is clear what changes are required.

**Comment by VladislavsPerkanuks** [comment]
@janishorsts Once again diff is in issue

File: parse/ast.go [hunk 2]
```diff
@@ -208,41 +209,32 @@ func (tp TextPattern) String() string {
...
-	Annotation Annotation // Optional: FunctionAnnotation, PrivateUseAnnotation, or ReservedAnnotation
+	Literal    Literal     // QuotedLiteral or UnquotedLiteral
+	Annotation Annotation  // Optional: FunctionAnnotation, PrivateUseAnnotation, or ReservedAnnotation
+	Attributes []Attribute // Optional
 }
 
 func (le LiteralExpression) String() string {
-	if le.Annotation == nil {
+	hasAnnotation := le.Annotation != nil
+	hasAttributes := len(le.Attributes) != 0
```
Comment (janishorsts) [inline parse/ast.go:226]:
```suggestion
	hasAttributes := len(le.Attributes) > 0
```

File: parse/ast.go [hunk 3]
```diff
@@ -208,41 +209,32 @@ func (tp TextPattern) String() string {
...
+
+	switch {
+	case !hasAnnotation && !hasAttributes: // Only literal
 		return fmt.Sprintf("{ %s }", le.Literal)
+	case hasAnnotation && !hasAttributes: // Literal + annotation
+		return fmt.Sprintf("{ %s %s }", le.Literal, le.Annotation)
+	case !hasAnnotation && hasAttributes: // Literal + attributes
+		return fmt.Sprintf("{ %s %s }", le.Literal, sliceToString(le.Attributes, " "))
+	default: // Literal + annotation + attributes
+		return fmt.Sprintf("{ %s %s %s }", le.Literal, le.Annotation, sliceToString(le.Attributes, " "))
```
Comment (janishorsts) [inline parse/ast.go:236]: Out of the scope of your PR. Curious.

Does empty string literal print `{  ` or `{ ||`?

Comment (VladislavsPerkanuks) [reply parse/ast.go:236]: latter

File: parse/parse.go [hunk 4]
```diff
@@ -242,26 +242,119 @@ func (p *parser) parsePatterns() ([]Pattern, error) {
...
+			continue
 		case itemText:
 			pattern = append(pattern, TextPattern(itm.val))
 		case itemExpressionOpen:
-			p.next() // skip opening brace
+			// HACK: Find if it's a markup or expression.
+			isMarkup := func() bool {
+				i := 1
+
+				for next := p.next(); ; next = p.next() {
```
Comment (janishorsts) [inline parse/parse.go:254]: Do you have to peek into the expression boundaries? E.g. this seems to go outside expression, you may find a closing markup in another expression.

Comment (janishorsts) [reply parse/parse.go:254]: Does simple `peekNonWS()` solve it?

```go
    if r := p.peekNonWS(); r ==  itemMarkupOpen || r == itemMarkupOpen {
       continue
    }
```
Markup follows immediately after the expression open.

```abnf
markup       = "{" [s] markup-open *(s attribute) [s] ["/"] "}"
             / "{" [s] markup-close *(s attribute) [s] "}"
markup-open  = "#" identifier *(s option)
markup-close = "/" identifier
```

Comment (VladislavsPerkanuks) [reply parse/parse.go:254]: need to test, but i tested many alternatives, this was the one that worked

Comment (VladislavsPerkanuks) [reply parse/parse.go:254]: would be nice if lexer gave itemMarkupStart\End, but it also needs some tweaks to make it work

File: parse/parse_test.go [hunk 5]
```diff
@@ -318,6 +278,41 @@ func TestParseSimpleMessage(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:  "markup",
+			input: `It is a {#button opt1=val1 @attr1=val1 } button { /button } this is a { #br /} something else`,
```
[the remark this answers is not in the document]
Comment (VladislavsPerkanuks) [reply parse/parse_test.go:283]: did not think about nesting...
Sure will do

**Review by DarjaLeonova — APPROVED** [review]

File: parse/lex.go [hunk 6]
```diff
@@ -732,3 +725,26 @@ func isDigit(r rune) bool {
 func isPrivateStart(r rune) bool {
 	return r == '^' || r == '&'
 }
+
+// isContent returns true if v is content character.
```
Comment (klavinsjanis) [inline parse/lex.go:729]: what does v represent in the comment?

File: parse/parse.go [hunk 7]
```diff
@@ -23,6 +23,16 @@ func (p *parser) next() item {
 	return p.items[p.pos]
 }
 
+func (p *parser) peekNonWS() item {
```
Comment (klavinsjanis) [inline parse/parse.go:26]: peekNonWs? Could you add a description?

File: parse/lex.go (hunk 6 above)
Comment (VladislavsPerkanuks) [reply parse/lex.go:729]: remains of old code, will fix

File: parse/parse.go (hunk 7 above)
Comment (VladislavsPerkanuks) [reply parse/parse.go:26]: peek next token which is not whitespace, should be pretty self explanatory, but i will add godoc

File: parse/ast.go [hunk 8]
```diff
@@ -704,6 +671,112 @@ func (v Variant) validate() error {
...
+	}
+
+	return nil
+}
+
+type MarkupType int
+
+const (
+	Unspecified MarkupType = iota
+	Open
```
Comment (klavinsjanis) [inline parse/ast.go:702]: Shouldn't these be prefixed with 'Markup'?

Comment (VladislavsPerkanuks) [reply parse/ast.go:702]: After a quick look, it seems prefixes are not necessary

File: parse/parse.go [hunk 9]
```diff
@@ -242,26 +252,104 @@ func (p *parser) parsePatterns() ([]Pattern, error) {
...
 		case itemError:
 			return nil, fmt.Errorf("got error token: '%s'", itm.val)
+		case itemWhitespace:
+			continue
 		case itemText:
 			pattern = append(pattern, TextPattern(itm.val))
 		case itemExpressionOpen:
-			p.next() // skip opening brace
+			// HACK: Find if it's a markup or expression, if it's markup, let the markup case handle it.
+			if next := p.peekNonWS(); next.typ == itemMarkupOpen || next.typ == itemMarkupClose {
```
Comment (janishorsts) [inline parse/parse.go:261]: alternative

```suggestion
			if typ := p.peekNonWS().typ; typ == itemMarkupOpen || typ == itemMarkupClose {
```

**Review by janishorsts — APPROVED** [review]

File: parse/parse.go (hunk 9 above)
Comment (VladislavsPerkanuks) [reply parse/parse.go:261]: yeah bit shorter, which is good

**Pull request merged**